Implement Promise interface

opedromiranda opened the pull request:
As mentioned in #20 .

```cache``` global removed. Using [```mem```](https://github.com/sindresorhus/mem) to cache return values.

[```execa```](https://github.com/sindresorhus/execa) being used instead of ```child_process ```.


#### getLocale regex

The regex was updated.

The system's ```locale``` command returns a set of lines like:
```
...
LC_COLLATE="en_GB.UTF-8"
LC_CTYPE="en_GB.UTF-8"
LC_MESSAGES="en_GB.UTF-8"
...
```

Each line is split by the ```=``` sign.

Given the string ```"en_GB.UTF-8"```,  replace would return ```"en_GB```.


```javascript
str = "\"en_GB.UTF-8\"";

str.replace(/[.:].*/, ''); // returns "en_GB
str.replace(/["]?(.*)[.:].*/, '$1'); // returns en_GB
```

sindresorhus commented:
Very nice! :)

I think the async implementation could be simplified further. Now that it's using promises we don't need the `.catch(() => fallback())` for each step, but can instead just catch the top-level returning promise. Same with the checks for empty values (`... || fallback();`).

sindresorhus commented on index.js:
```diff
@@ -28,101 +24,90 @@ function parseLocale(x) {
-		return;
-	}
+	const envLocale = getEnvLocale();
+	let locale;
```
This looks moot. It's not being used anywhere. It's more like the old design which is no longer necessary. Line 37 can also be inlined into line 38.

sindresorhus commented on index.js:
```diff
@@ -28,101 +24,90 @@ function parseLocale(x) {
-	};
+	const getAppleLocale = () => execa.stdout('defaults', ['read', '-g', 'AppleLocale'])
+			.then(stdout => {
+				locale = stdout.trim() || fallback();
```
`execa` trims the output for us, so `.trim()` is not needed.

sindresorhus commented on index.js:
```diff
@@ -28,101 +24,90 @@ function parseLocale(x) {
+	const envLocale = getEnvLocale();
+	let locale;
+
+	if (envLocale || !execa.sync || opts.spawn === false) {
```
`!execa.sync` is not needed as `execa.sync` always exists.

sindresorhus commented on test.js:
```diff
@@ -23,15 +23,19 @@ function restoreEnvVars(cache) {
 }
 
 test.cb('async', t => {
```
Use a normal test and async/await (AVA transpiles tests through Babel)

```js
test('async', async t => {
```

You can simplify the below code to not use `.then`s

Example: https://github.com/sindresorhus/pageres/blob/05457dca42a087cf43f1e0f0d195740dc32a6cba/test/test.js#L71-L76

sindresorhus commented:
To make the code more readable, I think we should split up all the check into top-level independent functions. That way the main code will just be concerned about calling the different checks and handling the fallback. For example, this could be an individual function: https://github.com/opedromiranda/os-locale/blob/0c35321fb8bf1ec3d4e1b0912449cc340064e68a/index.js#L83-L94

sindresorhus commented:
// @nexdrew @bcoe In case you have time to help review.

opedromiranda commented:
@sindresorhus Thanks for the feedback!
Related to the code splitting, how'd you feel about something like this:

```javascript
module.exports = mem(opts => {
	opts = opts || defaultOpts;

	const envLocale = getEnvLocale();

	if (envLocale || opts.spawn === false) {
		return Promise.resolve(getLocale(envLocale));
	}

	if (process.platform === 'win32') {
		return getWinLocaleAsync();
	}

	return getUnixLocaleAsync();
});

module.exports.sync = mem(opts => {
	opts = opts || defaultOpts;

	const envLocale = getEnvLocale();

	if (envLocale || opts.spawn === false) {
		return getLocale(envLocale);
	}

	if (process.platform === 'win32') {
		return getWinLocaleSync();
	}

	return getUnixLocaleSync();
});
```

sindresorhus commented:
@opedromiranda Yes. That's exactly what I was thinking.

> or maybe have only one fallback each per exported function

Yes

opedromiranda commented on index.js:
```diff
@@ -28,101 +24,90 @@ function parseLocale(x) {
 
 function getLocale(str) {
-	return (str && str.replace(/[.:].*/, '')) || fallback();
+	return (str && str.replace(/["]?(.*)[.:].*/, '$1')) || fallback();
```
I'm sorry but I don't know what do you mean by regression test in this scenario. Could you care to explain, please?

opedromiranda commented:
PR updated. :)

Switched the ```fallback``` function for a ```defaultLocale``` value.

Created the following functions: ```getAppleLocale```, ```getAppleLocaleSync```, ```getUnixLocale```, ```getUnixLocaleSync```, ```getWinLocale``` and ```getWinLocaleSync```.

```defaultLocale``` only used by  exported functions.

Previous regex restored now that ```"``` are removed in ```parseLocale```.

Removed ```execa.sync``` check.

Removed ```trim()```.

sindresorhus commented:
Thank you @opedromiranda. I appreciate you working on this. This was a very high quality pull request :)

![happy](https://cloud.githubusercontent.com/assets/170270/20348976/cc5e9f34-ac3a-11e6-81d9-410cb9663145.gif)
